Run conformance tests

philandstuff opened the pull request:
This runs the [conformance tests](https://github.com/openregister/conformance-test) as part of the build, both locally and in travis.

philandstuff committed:
wip: add conformance tests to build

philandstuff committed:
supporess https conformance tests in gradle build

In our early test environments, we don't have https set up.  This
suppresses https conformance tests in these environments.

om-sharma committed:
Recreate schema and load school data for conformance tests

om-sharma committed:
running psql queries quietly

philandstuff committed:
manage python environment using gradle

This avoids the user having to manage the virtual environment themselves
just to get the openregister-conformance test working.

philandstuff committed:
ensure python 3 is present on travis

philandstuff committed:
try more up-to-date python from ppa

The main `precise` repositories only go up to python 3.2.  However `pip`
now does not support 3.2 and requires >= 3.3.  We need `pip` in order to
install and run the conformance tests.

We can instead install a later python from a ppa.  Travis has a
whitelist of acceptable ppas to install software from.

The "deadsnakes" alias [1] refers to the whitelisted
ppa:fkrull/deadsnakes [2] which provides lots of python versions for
lots of ubuntu versions.  We use this to pull in python 3.5 instead.

[1]: https://github.com/travis-ci/apt-source-whitelist/blob/dce2f6b8fd2ccaf4fcd41dbe2f17c69e80d6d6a9/ubuntu.json#L72-L76
[2]: https://launchpad.net/~fkrull/+archive/ubuntu/deadsnakes

philandstuff committed:
need to also install pyvenv separately

although `venv` is part of the standard python 3 distribution (since
3.3), the deadsnakes packages split it into a separate package.

Also, they don't provide a `pyvenv` alias for the binary so it can only
be referred to as pyvenv-3.5.  I've changed the gradle file to use this
name.  I don't like this solution though, it feels brittle.  Longer term
I feel it'd be better to configure travis to create a `pyvenv` alias
somehow.

philandstuff commented:
rebased

michaelabenyohai commented on src/test/java/uk/gov/register/presentation/functional/testSupport/CleanDatabaseRule.java:
```diff
@@ -21,8 +19,8 @@ public CleanDatabaseRule(String pgUrl, String pgUser, String tableName) {
-                statement.execute("DROP TABLE IF EXISTS " + tableName);
-                statement.execute("CREATE TABLE " + tableName + " (SERIAL_NUMBER INTEGER PRIMARY KEY, ENTRY JSONB)");
+                statement.execute("DROP TABLE IF EXISTS ordered_entry_index");
+                statement.execute("CREATE TABLE ordered_entry_index (SERIAL_NUMBER INTEGER PRIMARY KEY, ENTRY JSONB, leaf_input varchar)");
```
Why is leaf_input included in the list of columns? Believe it was removed.

philandstuff committed:
fix bad merge

leaf_input was removed in #186, but because this branch edited the same
line it ended up trying to reintroduce it.

michaelabenyohai commented on create_schema.sh:
```diff
@@ -0,0 +1,23 @@
+
+db_name=$1
+
+createdb $db_name
```
Do we want to check whether the database exists before creating it? To avoid the below error.

"createdb: database creation failed: ERROR:  database "ft_presentation" already exists"

michaelabenyohai commented on run-application.sh:
```diff
@@ -2,19 +2,7 @@
-psql $db_name -U postgres -c "insert into total_records (count) select 0 where not exists (select count from total_records)"
-
-psql $db_name -U postgres -c "CREATE TABLE IF NOT EXISTS ordered_entry_index(serial_number INTEGER PRIMARY KEY, entry JSONB)"
-psql $db_name -U postgres -c "ALTER TABLE ordered_entry_index drop column leaf_input"
```
Maybe this should be ALTER TABLE ordered_entry_index drop column **if exists** leaf_input. Would avoid the below warning.

"NOTICE:  column "leaf_input" of relation "ordered_entry_index" does not exist, skipping"

philandstuff replied:
already done :point_up_2: :smile:

philandstuff committed:
Don't try to createdb in create_schema.sh

I tested this functionality by dropping my `ft_presentation` db and
trying to run the build.  The app failed to start, because the
ft_presentation db didn't exist.

The problem is that `create_schema.sh` is run after the app is started,
so creating the db at this point is too late to allow the app to run.

We already have the `./go` script to create this db, which is referenced
from the README, so I think we can just remove the createdb call from
create_schema.sh.

philandstuff commented:
responded to comments, thanks for feedback
